Adds 'Required' label for merchant on workspaces & updates P2P requirements for worskpaces

cdanwards opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$https://github.com/Expensify/App/issues/36169
$https://github.com/Expensify/App/issues/35012
PROPOSAL:

In this PR we are adding the `Required` label to the `Merchant` category whenever making money requests within Workspaces. We're also updating the way in which it's determined if a request belongs to a workspace in some instances.

### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

#### Testing for creating a money request in a workspace (Required should appear)
- [ ] Verify that no errors appear in the JS console
- [ ] Navigate to a workspace
- [ ] Within the creation menu, select "Request Money"
- [ ] Enter in an amount
- [ ] Select "Next"
- [ ] Verify that the Merchant Menu Item has a "Required" as a right label

#### Testing for creating a p2p money request (Required should not appear)
- [ ] Verify that no errors appear in the JS console
- [ ] Within the global creation menu, select "Request Money"
- [ ] Enter in an amount
- [ ] Select "Next"
- [ ] Select a peer for a money request
- [ ] Select "Next"
- [ ] Expand to see all menu items
- [ ] Verify that the Merchant Menu Item does not have a "Required" as a right label

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

#### Same as Tests

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

#### Same as Tests

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<img width="300" src="https://github.com/Expensify/App/assets/8878532/23eb961e-4665-4590-95d9-8ce641683c32" />

</details>

<details>
<summary>Android: mWeb Chrome</summary>

<img width="300" src="https://github.com/Expensify/App/assets/8878532/f5f6ca1f-38ed-4ef3-b72c-5a3dbf5331da" />

</details>

<details>
<summary>iOS: Native</summary>

<img width="300" src="https://github.com/Expensify/App/assets/8878532/f66d7658-7891-4280-9081-f0feb823b6bb" />

</details>

<details>
<summary>iOS: mWeb Safari</summary>

<img width="300" src="https://github.com/Expensify/App/assets/8878532/4dffc0bd-aeef-483e-99ce-0766e0682cac" />

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<img width="600" src="https://github.com/Expensify/App/assets/8878532/128372a8-22f1-4d93-a64d-6b2b7e1b3094" />

</details>

<details>
<summary>MacOS: Desktop</summary>

<img width="600" src="https://github.com/Expensify/App/assets/8878532/1e6b2515-df96-4327-820b-1345ec7f0016" />

</details>

cdanwards committed:
Update MoneyRequestConfirmationList component

cdanwards committed:
Merge branch 'lindboe/violations/money-request-updates' of github.com:infinitered/ExpensifyApp into violations/cdanwards/merchant-required

cdanwards committed:
Refactor MoneyRequestConfirmationList and MoneyTemporaryForRefactorRequestConfirmationList***

***Update IOU.ts to handle policy requirements

cdanwards committed:
Merge branch 'lindboe/violations/money-request-updates' of github.com:infinitered/ExpensifyApp into violations/cdanwards/merchant-required

cdanwards committed:
Merge branch 'main' of github.com:Expensify/App into violations/cdanwards/merchant-required

cdanwards committed:
Refactor rightLabel in MoneyTemporaryForRefactorRequestConfirmationList.js

cdanwards committed:
Merge branch 'main' of github.com:Expensify/App into violations/cdanwards/merchant-required

cdanwards commented on src/components/MoneyRequestConfirmationList.js:
```diff
@@ -755,6 +755,7 @@ function MoneyRequestConfirmationList(props) {
                                     ? translate('common.error.enterMerchant')
                                     : ''
                             }
+                            rightLabel={canUseViolations && PolicyUtils.isPaidGroupPolicy(props.policy) ? translate('common.required') : ''}
```
Adding this in here to keep consistency between this and the temporary file.

cead22 replied:
Let's remove `canUseViolations` from here
```suggestion
                            rightLabel={PolicyUtils.isPaidGroupPolicy(props.policy) ? translate('common.required') : ''}
```

cdanwards commented on src/components/MoneyTemporaryForRefactorRequestConfirmationList.js:
```diff
@@ -729,6 +729,7 @@ function MoneyTemporaryForRefactorRequestConfirmationList({
                     interactive={!isReadOnly}
                     brickRoadIndicator={merchantError ? CONST.BRICK_ROAD_INDICATOR_STATUS.ERROR : ''}
                     error={merchantError ? translate('common.error.fieldRequired') : ''}
+                    rightLabel={canUseViolations && PolicyUtils.isPaidGroupPolicy(policy) ? translate('common.required') : ''}
```
Unsure if the `PolicyUtils.isPaidGroupPolicy(policy)` is redundant. `isMerchantRequired` checks for if the policy is of type `CONST.REPORT.CHAT_TYPE.POLICY_EXPENSE_CHAT` which might be a level higher than paid group policy. Is that the case @cead22 ?

cead22 replied:
I think so, can you test to confirm? We can get rid of `canUseViolations` here as well

cdanwards commented:
@cead22 this is now ready for review!

cead22 commented on src/libs/actions/IOU.ts:
```diff
@@ -675,11 +675,11 @@ function buildOnyxDataForMoneyRequest(
     }
 
-    const violationsOnyxData = ViolationsUtils.getViolationsOnyxData(transaction, [], !!policy.requiresTag, policyTags ?? {}, !!policy.requiresCategory, policyCategories ?? {});
+    const violationsOnyxData = ViolationsUtils.getViolationsOnyxData(transaction, [], !!policy?.requiresTag, policyTags ?? {}, !!policy?.requiresCategory, policyCategories ?? {});
```
Why do we need this change? `policy` should be set here, no?

cead22 commented on src/libs/actions/IOU.ts:
```diff
@@ -675,11 +675,11 @@ function buildOnyxDataForMoneyRequest(
 
     // Policy won't be set for P2P cases for which we don't need to compute violations
-    if (!policy?.id) {
+    if (policy && PolicyUtils.isPaidGroupPolicy(policy)) {
```
Is this condition backwards?
```suggestion
    if (!policy || !PolicyUtils.isPaidGroupPolicy(policy)) {
```

cead22 reviewed (changes requested):
Please add manual tests for the case of p2p money requests (where the policy isn't set)

cdanwards committed:
Fix condition in buildOnyxDataForMoneyRequest function

cdanwards committed:
Refactor code to use translated label for required field and other updates

cdanwards commented:
@cead22 updated with a p2p test case. Also updated that check like you described. The optional chaining was only necessary with the check the way I had it, so I removed those as well.

cead22 commented on src/components/MoneyTemporaryForRefactorRequestConfirmationList.js:
```diff
@@ -729,6 +729,7 @@ function MoneyTemporaryForRefactorRequestConfirmationList({
                     interactive={!isReadOnly}
                     brickRoadIndicator={merchantError ? CONST.BRICK_ROAD_INDICATOR_STATUS.ERROR : ''}
                     error={merchantError ? translate('common.error.fieldRequired') : ''}
+                    rightLabel={translate('common.required')}
```
Don't we need the `PolicyUtils.isPaidGroupPolicy` check here?

cead22 commented on src/components/MoneyRequestConfirmationList.js:
```diff
@@ -755,6 +755,7 @@ function MoneyRequestConfirmationList(props) {
                                     ? translate('common.error.enterMerchant')
                                     : ''
                             }
+                            rightLabel={translate('common.required')}
```
Don't we need the `PolicyUtils.isPaidGroupPolicy` check here?

cdanwards replied:
@cead22  My understanding is that this `Merchant` `MenuItem` is only shown whenever `isPolicyExpenseChat`, where in the case of P2P its a little further down.

This `isMerchantRequired` ultimately is reliant up on the report chat type being `policyExpenseChat` [here is the code link](https://github.com/Expensify/App/blob/784aa27a848867d85ff5f927d4267e2b84af747f/src/libs/ReportUtils.ts#L777) (the `isPaidGroupPolicy` check is a little further down. Are we looking expense chat reports that are exclusively from Control or Collect policies?

cead22 replied:
Ah, yes, I see that `isMerchantRequired` for the case of `src/components/MoneyTemporaryForRefactorRequestConfirmationList.js`. For this one we have `shouldShowMerchant` and I don't see it relying on `policyExpenseChat`, but if it's showing for workspace requests and not showing for p2p requests, we're good

cdanwards replied:
Yeah the dependency of that check goes `isMerchantRequired` > `isPolicyExpenseCheck` > `ReportUtils.isPolicyExpenseChat` (in the parent component). This file isn't up to date it looks like but the temporary other file is!

cead22 replied:
Nice 👌 

> This file isn't up to date it looks like but the temporary other file is!

Do we need to update this file, or what do you mean here?

cdanwards replied:
I've updated this file, the original `MoneyRequestConfirmationList` file, whenever I've made a change in the temporary `MoneyTemporaryForRefactorRequestConfirmationList` file, but it looks like the temporary file is now being treated as the source of truth for this component and now the original file is out of date. We don't need to make any more changes!

cead22 commented on src/libs/actions/IOU.ts:
```diff
@@ -675,7 +675,7 @@ function buildOnyxDataForMoneyRequest(
     ];
 
     // Policy won't be set for P2P cases for which we don't need to compute violations
```
Sorry I missed this before, but this is now outdated
```suggestion
    // We don't need to compute violations unless we're on a paid policy
```

cead22 commented on src/libs/actions/IOU.ts:
```diff
@@ -2650,7 +2650,7 @@ function editRegularMoneyRequest(
     ];
 
     // Add transaction violations if there is a policy and updated transaaction
```
```suggestion
    // Add transaction violations if we have a paid policy and an updated transaction
```

cdanwards committed:
Update src/libs/actions/IOU.ts

Co-authored-by: Carlos Alvarez <cead22@gmail.com>

cdanwards committed:
Update src/libs/actions/IOU.ts

Co-authored-by: Carlos Alvarez <cead22@gmail.com>

cdanwards committed:
Merge branch 'main' of github.com:Expensify/App into violations/cdanwards/merchant-required

cdanwards committed:
Merge branch 'main' of github.com:Expensify/App into violations/cdanwards/merchant-required

cdanwards committed:
Add right label for paid group policy

cdanwards commented:
`MoneyTemporaryForRefactorRequestConfirmationList` was refactored so I now needed to use the `PolicyUtils.isPaidGroupPolicy` util.

cdanwards committed:
Update src/components/MoneyTemporaryForRefactorRequestConfirmationList.js

Co-authored-by: Aimane Chnaif <96077027+aimane-chnaif@users.noreply.github.com>

cdanwards commented on src/libs/actions/IOU.ts:
```diff
@@ -674,8 +674,8 @@ function buildOnyxDataForMoneyRequest(
-    // Policy won't be set for P2P cases for which we don't need to compute violations
-    if (!policy?.id) {
+    // We don't need to compute violations unless we're on a paid policy
+    if (!policy || !PolicyUtils.isPaidGroupPolicy(policy)) {
```
This is to satisfy TS. `isPaidGroupPolicy` expects `EmptyObject` or `OnyxEntry<Policy>` so it throws a type error for undefined.

cdanwards commented:
@aimane-chnaif I went back and looked at the issue and you are correct that it should be for all workspaces. I'll change which check is made.

cdanwards committed:
Refactor rightLabel in MoneyTemporaryForRefactorRequestConfirmationList.js

cdanwards committed:
Merge branch 'main' of github.com:Expensify/App into violations/cdanwards/merchant-required

cead22 commented on src/libs/actions/IOU.ts:
```diff
@@ -1226,15 +1226,15 @@ function getUpdateMoneyRequestParams(
                 updatedTransaction,
                 currentTransactionViolations,
-                !!policy.requiresTag,
+                !!policy?.requiresTag,
```
Do we need the `?`s here and below? if `PolicyUtils.isPaidGroupPolicy(policy)` returns true then policy is an object right?

Same question for the ViolationsUtils.getViolationsOnyxData call below on line 2670

lindboe replied:
According to TypeScript we would (and generally TS linters will error if you're optionally chaining unnecessarily). I think if we add a check that policy exists in the if statement above, TS will narrow the type in that scope and know it exists, but from just the `isPaidGroupPolicy` call, TS can't do that narrowing. (It could if `isPaidGroupPolicy` was a type guard that narrowed a nullable general policy type to a paid policy type, or just a nullable policy to a non-null policy, but it's just a predicate).

cdanwards replied:
Looks like in this call, the type of `policy` is `OnyxEntry<OnyxTypes.Policy>`, which the `OnyxEntry` wrapper types the value as `TOnyxValue | null` and so the optional chaining is to satisfy the possibility of it resolving as `null`.

Further down in the other `ViolationsUtils.getViolationsOnyxData` call the `policy` is typed as just `OnyxTypes.Policy` so there is no need for the chaining down there.

cead22 replied:
You mean like this? If so, that sounds like a good improvement to me
```
if (policy && PolicyUtils.isPaidGroupPolicy(policy) && updatedTransaction) {
```

cdanwards committed:
Fix policy check in getUpdateMoneyRequestParams function

cead22 commented on src/libs/actions/IOU.ts:
```diff
@@ -2664,8 +2664,8 @@ function editRegularMoneyRequest(
-    // Add transaction violations if there is a policy and updated transaaction
-    if (policy?.id && updatedTransaction) {
+    // Add transaction violations if we have a paid policy and an updated transaction
+    if (PolicyUtils.isPaidGroupPolicy(policy) && updatedTransaction) {
```
Let's update this one too

cdanwards committed:
Fix conditional statement in editRegularMoneyRequest function

cdanwards committed:
Removed prop from MoneyRequestConfirmationList

cdanwards committed:
Merge branch 'main' of github.com:Expensify/App into violations/cdanwards/merchant-required

cdanwards commented on src/components/MoneyTemporaryForRefactorRequestConfirmationList.js:
```diff
@@ -720,6 +720,7 @@ function MoneyTemporaryForRefactorRequestConfirmationList({
                     interactive={!isReadOnly}
                     brickRoadIndicator={merchantError ? CONST.BRICK_ROAD_INDICATOR_STATUS.ERROR : ''}
                     error={merchantError ? translate('common.error.fieldRequired') : ''}
+                    rightLabel={isMerchantRequired ? translate('common.required') : ''}
```
Yeah, this is just following the design precedent set by the other required tags.

cead22 replied:
There was a bug created to remove the Required label from the category and tag fields if they have a values set. I can include the merchant field in the scope of that bug

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/7966219100) to staging by https://github.com/cead22 in version: 1.4.43-0 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/8055338936) to production by https://github.com/puneetlath in version: 1.4.43-20 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅
